src/charm.py

```diff
@@ -642,580 +92,247 @@ def k8s_auth(self) -> None:
+        self._update_ingress()
 
-        Args:
-            event: The event that fires this method.
+    def _on_data_removed(self, _: Any) -> None:
+        """Handle the data-removed event."""
+        self._update_ingress()
 
-        Raises:
-            An error if there are insufficient permissions to delete an ingress.
-            ConflictingAnnotationsError: if there are conflicting annotations.
-            ConflictingRoutesError: if there are conflicting routes.
-            InvalidHostnameError: if there is an invalid hostname.
-            InvalidBackendProtocolError: an invalid backend protocol.
-        """
-        conf_or_rel = _ConfigOrRelation(self.model, {}, event.relation, self._multiple_relations)
-        if self.unit.is_leader() and conf_or_rel._ingress_name:
-            try:
-                # NOTE: _define_ingresses will recreate the Kubernetes Ingress Resources based
-                # on the existing relations, and remove any resources that are no longer needed
-                # (they were needed by the event relation).
-                self._define_ingresses(excluded_relation=event.relation)
-                self._delete_unused_resources()
-            except kubernetes.client.exceptions.ApiException as exception:
-                if exception.status == 403:
-                    LOGGER.error(
-                        "Insufficient permissions to delete the k8s ingress resource, "
-                        "will request `juju trust` to be run"
-                    )
-                    juju_trust_cmd = f"juju trust {self.app.name} --scope=cluster"
-                    self.unit.status = BlockedStatus(
-                        f"Insufficient permissions, try: `{juju_trust_cmd}`"
-                    )
-                    return
-                raise
-            except ConflictingAnnotationsError:
-                self.unit.status = BlockedStatus(
-                    "Conflicting annotations from relations. Run juju debug-log for details. "
-                    "Set manually via juju config."
-                )
-                return
-            except ConflictingRoutesError:
-                self.unit.status = BlockedStatus(
-                    "Duplicate route found; cannot add ingress. Run juju debug-log for details."
-                )
-                return
-            except InvalidHostnameError:
-                self.unit.status = BlockedStatus(INVALID_HOSTNAME_MSG)
-                return
-            except InvalidBackendProtocolError:
-                self.unit.status = BlockedStatus(INVALID_BACKEND_PROTOCOL_MSG)
-                return
-        self.unit.status = ActiveStatus()
+    def _on_nginx_route_available(self, _: Any) -> None:
+        """Handle the nginx-route-available event."""
+        self._update_ingress()
+
+    def _on_nginx_route_broken(self, _: Any) -> None:
+        """Handle the nginx-route-broken event."""
+        self._update_ingress()
```

Comment:
I was wondering if in such cases where all events lead to reconcile, we can implement the [reconcile pattern](https://discourse.charmhub.io/t/eliminating-duplication-across-charms-by-using-a-reusable-reconcile-function/11304), I don't think it is currently approved internally within the team but just leaving it here quietly... 😂

Reply:
Yes, I have received feedback suggesting a dedicated function handler with different names for each event...